amigo-lib/src/main/java/me/ele/amigo/AmigoDirs.java

```diff
@@ -4,94 +4,139 @@
                 return;
             }
+
             amigoDir = new File(context.getFilesDir(), AMIGO_FOLDER_NAME);
             FileUtils.mkdirChecked(amigoDir);
 
-            dexDir = new File(amigoDir, AMIGO_DEX_FOLDER_NAME);
-            libDir = new File(amigoDir, AMIGO_LIB_FOLDER_NAME);
-            FileUtils.mkdirChecked(dexDir);
-            FileUtils.mkdirChecked(libDir);
-
-            File cacheDir = new File(applicationInfo.dataDir, CODE_CACHE_NAME);
+            cacheDir = new File(applicationInfo.dataDir, CODE_CACHE_NAME);
             FileUtils.mkdirChecked(cacheDir);
-            optDir = new File(cacheDir, CODE_CACHE_AMIGO_DEX_FOLDER_NAME);
-            FileUtils.mkdirChecked(optDir);
         } catch (Exception e) {
             throw new RuntimeException("Initiate amigo files failed (" + e.getMessage() + ").");
         }
     }
 
+    private void ensurePatchDirs(String checksum) {
+        try {
+            ensureAmigoDir();
+            File patchDir = new File(amigoDir, checksum);
+            if (!patchDir.exists()) {
+                FileUtils.mkdirChecked(patchDir);
+            }
+
+            File dexDir = new File(patchDir, AMIGO_DEX_FOLDER_NAME);
+            if (!dexDir.exists()) {
+                FileUtils.mkdirChecked(dexDir);
+            }
+            dexDirs.put(checksum, dexDir);
+
+            File libDir = new File(patchDir, AMIGO_LIB_FOLDER_NAME);
+            if (!libDir.exists()) {
+                FileUtils.mkdirChecked(libDir);
+            }
+            libDirs.put(checksum, libDir);
+
+            File patchCacheDir = new File(cacheDir, checksum);
+            if (!patchCacheDir.exists()) {
+                FileUtils.mkdirChecked(patchCacheDir);
+            }
+            File optDir = new File(patchCacheDir, CODE_CACHE_AMIGO_DEX_FOLDER_NAME);
+            if (!optDir.exists()) {
+                FileUtils.mkdirChecked(optDir);
+            }
+            optDirs.put(checksum, optDir);
+        } catch (Exception e) {
+            throw new RuntimeException("Initiate amigo files for patch apk: " + checksum + " failed (" + e.getMessage() + ").");
+        }
+    }
+
     public File amigoDir() {
+        ensureAmigoDir();
         return amigoDir;
     }
 
-    public File dexOptDir() {
-        return optDir;
+    public File dexOptDir(String checksum) {
+        ensurePatchDirs(checksum);
+        checkInitState(optDirs, checksum);
+        return optDirs.get(checksum);
     }
 
-    public File dexDir() {
-        return dexDir;
+    public File dexDir(String checksum) {
+        ensurePatchDirs(checksum);
+        checkInitState(dexDirs, checksum);
+        return dexDirs.get(checksum);
     }
 
-    public File libDir() {
-        return libDir;
+    public File libDir(String checksum) {
+        ensurePatchDirs(checksum);
+        checkInitState(libDirs, checksum);
+        return libDirs.get(checksum);
+    }
+
+    private void checkInitState(Map<String, File> map, String checksum) {
```

Comment:
since #ensurePatchDirs would throw an exception if any error occurred during creating those files,  I don't think this check would be that necessary.